### ISSUE TITLE
VariableDeclarationExpr should not be an Expression

### ISSUE THREAD

**Issue by ShiboTanany** [javaparser/javaparser#1905]
this code for analysis 
```java
method.getBody().get().getStatements().forEach(existedStmt -> {
						ExpressionStmt express = (ExpressionStmt) existedStmt;
						String newStmt = existedStmt.toString().replace("this", beanClass.toLowerCase()).replace(";", "");
								express.setExpression(newStmt);});
```
this code that will be parsed 
```java
final Long changeID = getDao().changeLog(this.x, change);
```

**Comment by matozoid** [comment]
Thanks, I'll take a look at what's happening here when I find some time.

**Comment by malteskoruppa** [comment]
Interesting!

Simplified example:

```java
Statement stmt = JavaParser.parseStatement("final int i = 42;"); // works
VariableDeclarationExpr expr1 = JavaParser.parseVariableDeclarationExpr("final int i = 42"); // works
Expression expr2 = JavaParser.parseExpression("final int i = 42"); // fails with the exception mentioned by OP
```

So you can't parse a variable declaration expression by using `parseExpression(String)`. JavaParser needs more help, you have to use `parseVariableDeclarationExpr(String)`.

The method `NodeWithExpression#setExpression(String)` called by the OP uses `JavaParser#parseExpression(String)` to convert the given string to an expression, so if you give it a variable declaration expression, this exception occurs.

**Comment by ShiboTanany** [comment]
@malteskoruppa  excuse me i need as example to setexpression with replace this to another i haven't way to do it without setexpression

**Comment by matozoid** [comment]
Well, the main problem is trying to do string operations on an AST instead of manipulating the nodes. Ignoring that, a quick would be to use `existedStmt.replace` with the `parseStatement` from @malteskoruppa.

Do read the book: https://leanpub.com/javaparservisited

**Comment by malteskoruppa** [comment]
@ShiboTanany Sorry if my post was unclear. It wasn't meant as a solution for your problem. Rather, it was meant to point out what the problem is internally.

I completely agree with @matozoid: You shouldn't manipulate strings. You should manipulate the nodes directly. Also see: http://wiki.c2.com/?StringlyTyped.

And yes, as a quick fix, you can use:

```java
ExpressionStmt express = JavaParser.parseStatement(existedStmt.toString().replace("this", beanClass.toLowerCase())).asExpressionStmt();
```

instead of

```java
ExpressionStmt express = (ExpressionStmt) existedStmt;
String newStmt = existedStmt.toString().replace("this", beanClass.toLowerCase()).replace(";", "");
express.setExpression(newStmt);
```

But, meh. You should really consider manipulating the nodes directly instead of taking a detour via strings.

**Comment by malteskoruppa** [comment]
@matozoid Nevertheless, wouldn't you consider this a bug?

The method `NodeWithExpression#setExpression(String)` suggests that you can pass it any string that is a valid expression. But if you pass it a string that is a _variable declaration_ expression, it just crashes. Even though a variable declaration expression is a valid expression. It's a somewhat surprising behavior.

More generally, it's a bit weird that `parseExpression(String)` crashes for a particular kind of (perfectly valid) expressions. Shouldn't the Javadoc of this method at least mention that the method cannot be used for variable declaration expressions, and that one should use `parseVariableDeclarationExpr(String)` instead in that case?

**Comment by matozoid** [comment]
Yep, there are two issues even:
- the documentation issue
- setExpression(String) cannot take a variable declaration expression.

It's like "in some places where you can have an expression, you can alternatively put a variable declaration." The JLS doesn't call them expressions, it seems. I guess this is a bit of creativity from aancient times to keep some type safety here (instead of typing all positions where you can have an expr or a variable decl a `Node`).

Sooo point two actually makes sense: a variable declaration is not an expression, so you can't set it with setExpression. The new issue is:
- VariableDeclarationExpr is not really an expr, but it is named that way, it is extended that way, and it is parsed inside an `ExpressionStatement` where it should not be.

--> https://docs.oracle.com/javase/specs/jls/se9/html/jls-14.html#jls-14.4

And since it's by design, this is not a bug but a possible improvement :-)

**Comment by malteskoruppa** [comment]
You are right of course. :-) The JLS is quite right to not call variable declarations "expressions"; they aren't. So they shouldn't be expressions in JavaParser either. That would be an improvement indeed!

Actually, looking at the syntax, a `LocalVariableDeclaration` can _only_ occur in a `LocalVariableDeclarationStatement`, and a `LocalVariableDeclarationStmt` can _only_ occur in a `BlockStatement`.
-> https://docs.oracle.com/javase/specs/jls/se9/html/jls-19.html

So, the most JLS-compatible way for JavaParser may be to get rid of `VariableDeclarationExpr` and introduce a `LocalVariableDeclarationStmt` instead. This statement could then contain a list of `VariableDeclarator`s (just as `FieldDeclaration` also contains a list of `VariableDeclarator`s). :-)

**Comment by matozoid** [comment]
You can find ample examples in the issue tracker where I was absolutely wrong ;-)

But what would you do about the variable declarations in `for(int i=1;;)` and `try(R r=reader()){}`?

**Comment by malteskoruppa** [comment]
Ah, yes, I see. :-)

Well, err, I seem to have missed that a `LocalVariableDeclaration` can _also_ occur in the init-part of a for-loop, even though the link I posted above clearly says so and I did Ctrl+F for "LocalVariableDeclaration" on that page, yet I missed it. ;-) As far as try-with-resource statements are concerned, the JLS declares its very own type for that (`ResourceSpecification`).

On the one hand I always tend to think "It's best to do things exactly as they are in the JLS", but, on the other, it seems like if JavaParser declared a type for each and every nonterminal symbol in the Java grammar, it would make the AST overly complicated. So, some simplification is desirable. And simplifications then lead to problems such as the one discussed here. Bummer. :-)

Well, for this particular case, hmm... maybe introduce an interface `ExpressionOrVariableDeclaration` or something like that, which could be used in the init-part of for-loops and the resources-part of try-with-resource-statements? Or maybe better (instead of misusing interfaces) just use a `Node` instead of an `Expression` in these situations, as you noted earlier?

**Comment by matozoid** [comment]
Yes, the usual way would be to just be less specific and use `Node`. That's why I agree that JP is lacking here, but not too eager to change anything because: it's a big breaking change, it reduces type safety, and it has led to only 1 issue (this one) where the real answer is: "please don't do it like you're doing it now."
